### PR TITLE
Memory profiling

### PR DESCRIPTION
**Pull request by ilia-cher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37587 RecordFunction in Dispatcher
* #37639 Use TensorMethods.cpp
* #38453 Fixes for profiling JIT code
* **#37775 Memory profiling**

Summary:
Adding memory usage into profiler table output

Test Plan:
```
BUILD_BINARY=1 USE_BLAS=MKL USE_MKLDNN=1 USE_CUDA=1 python setup.py develop install
```

```
$ python benchmarks/profiler_benchmark/resnet_memory_profiler.py
output: https://gist.github.com/ilia-cher/3f37d54c3b2afb24d6776858e6860f69
```

```
$ python test/test_autograd.py TestAutograd.test_memory_profiler
Couldn't download test skip set, leaving all tests enabled...
Running CPU test
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                         Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CPU Mem          Self CPU Mem     CUDA Mem         Self CUDA Mem    Number of Calls  Input Shapes
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
test_user_scope_alloc        60.58%           105.892us        93.42%           163.285us        163.285us        800 b            0 b              0 b              0 b              1                []
rand                         10.53%           18.405us         32.83%           57.393us         57.393us         800 b            0 b              0 b              0 b              1                []
empty                        1.77%            3.092us          1.77%            3.092us          3.092us          800 b            800 b            0 b              0 b              1                []
uniform_                     19.64%           34.325us         20.54%           35.896us         35.896us         0 b              0 b              0 b              0 b              1                [[10, 10]]
is_complex                   0.90%            1.571us          0.90%            1.571us          1.571us          0 b              0 b              0 b              0 b              1                [[10, 10]]
test_user_scope_dealloc      6.58%            11.508us         6.58%            11.508us         11.508us         -800 b           -800 b           0 b              0 b              1                []
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 174.793us

Running CUDA test
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                         Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CPU Mem          Self CPU Mem     CUDA Mem         Self CUDA Mem    Number of Calls  Input Shapes
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
test_user_scope_alloc        29.37%           86.836us         93.05%           275.143us        275.143us        0 b              -800 b           1.00 Kb          0 b              1                []
to                           7.42%            21.939us         51.31%           151.703us        151.703us        0 b              0 b              1.00 Kb          0 b              1                [[10, 10]]
empty_strided                6.19%            18.295us         6.19%            18.295us         18.295us         0 b              0 b              1.00 Kb          1.00 Kb          1                []
rand                         4.50%            13.316us         12.38%           36.604us         36.604us         800 b            0 b              0 b              0 b              1                []
empty                        0.83%            2.456us          0.83%            2.456us          2.456us          800 b            800 b            0 b              0 b              1                []
uniform_                     6.44%            19.044us         7.05%            20.832us         20.832us         0 b              0 b              0 b              0 b              1                [[10, 10]]
is_complex                   0.60%            1.788us          0.60%            1.788us          1.788us          0 b              0 b              0 b              0 b              1                [[10, 10]]
copy_                        37.70%           111.469us        37.70%           111.469us        111.469us        0 b              0 b              0 b              0 b              1                [[10, 10], [10, 10]]
test_user_scope_dealloc      6.95%            20.544us         6.95%            20.544us         20.544us         0 b              0 b              -1.00 Kb         -1.00 Kb         1                []
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 295.687us

Running MKLDNN test
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Name                         Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CPU Mem          Self CPU Mem     CUDA Mem         Self CUDA Mem    Number of Calls  Input Shapes
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
test_user_scope_alloc        34.23%           43.503us         88.57%           112.550us        112.550us        400 b            -400 b           0 b              0 b              1                []
rand                         8.00%            10.167us         18.34%           23.302us         23.302us         400 b            0 b              0 b              0 b              1                []
empty                        2.22%            2.815us          2.22%            2.815us          2.815us          400 b            400 b            0 b              0 b              1                []
to_mkldnn                    35.16%           44.675us         36.00%           45.745us         45.745us         400 b            400 b            0 b              0 b              1                [[10, 10]]
uniform_                     7.24%            9.198us          8.12%            10.320us         10.320us         0 b              0 b              0 b              0 b              1                [[10, 10]]
is_complex                   0.88%            1.122us          0.88%            1.122us          1.122us          0 b              0 b              0 b              0 b              1                [[10, 10]]
contiguous                   0.84%            1.070us          0.84%            1.070us          1.070us          0 b              0 b              0 b              0 b              1                [[10, 10]]
test_user_scope_dealloc      11.43%           14.525us         11.43%           14.525us         14.525us         -400 b           -400 b           0 b              0 b              1                []
---------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  -----------------------------------
Self CPU time total: 127.075us

.
----------------------------------------------------------------------
Ran 1 test in 1.571s

OK
```

Differential Revision: [D21384248](https://our.internmc.facebook.com/intern/diff/D21384248)

### CODE REVIEW

**Review by dzhulgakov — CHANGES_REQUESTED** [review]
What about testing? Can you write unittests? (including enabling profiler in the middle of the allocation)

I'd also double-triple check that memory allocation tracking works well with non-standard allocators, e.g. if one does MKLDNN allocations (with to_mkldnn) or some of the internal ones (huge pages)

File: c10/core/Allocator.h [hunk 1]
```diff
@@ -219,4 +220,19 @@ struct AllocatorRegisterer {
   static AllocatorRegisterer<t> g_allocator_d(f); \
   }
 
+// An interface for reporting thread local memory usage
+// per device
+struct C10_API MemoryReportingInfoBase : public c10::DebugInfoBase {
+  MemoryReportingInfoBase() {}
```
Comment (dzhulgakov) [inline c10/core/Allocator.h:226]: as discussed - move it to .cpp file to avoid potentially duplicated symbols

**Review by dzhulgakov — APPROVED** [review]
Great! I think it's good to go! (and nice catch on overlapping ranges)

There's also some allocator stuff left in TH, but I'm not sure what it is and I didn't trace where it gets called:
https://github.com/pytorch/pytorch/blob/40265e2d663cc0027cffa6e80ee1ec67d467ca00/aten/src/TH/THAllocator.cpp

**Comment by ilia-cher** [comment]
from what I got (thanks @ezyang) this is a special allocator used to allocate tensors in shared memory space used for inter-process communication; I guess we can add memory reporting to there too

**Comment by ezyang** [comment]
This broke ROCm tests:

```

01:17:35 ======================================================================
01:17:35 FAIL: test_memory_profiler (__main__.TestAutograd)
01:17:35 ----------------------------------------------------------------------
01:17:35 Traceback (most recent call last):
01:17:35   File "test_autograd.py", line 2940, in test_memory_profiler
01:17:35     "test_user_scope_dealloc",
01:17:35   File "test_autograd.py", line 2897, in check_metrics
01:17:35     self.assertTrue(stat_metrics[alloc_fn] > 0)
01:17:35 AssertionError: False is not true
01:17:35 
01:17:35 ----------------------------------------------------------------------
```

**Comment by ezyang** [comment]
Was failing on https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py3.6-clang7-rocmdeb-ubuntu16.04-test2/28393/console

cc @jeffdaily

**Comment by jeffdaily** [comment]
Thank you for the notification.  In the future, if a new test for a new feature is breaking ROCm CI, can we have the developer(s) add the `skipIfRocm` decorator and tag me to look into it?  This would help with our CI stability.  Ideally, it would be fixed prior to merging the PR, but I understand at present it's a tough request.

**Comment by jeffdaily** [comment]
cc @ezyang https://github.com/pytorch/pytorch/pull/38790

**Comment by ilia-cher** [comment]
I'm landing the fix https://github.com/pytorch/pytorch/pull/38795

**Comment by ilia-cher** [comment]
have you noticed though that py3.6-clang7-rocmdeb-ubuntu16.04-test2 was broken on trunk at least since May 16 
https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master ?

**Comment by albanD** [comment]
I don't think it was continuously broken, it was fixed on the 19th, then broken again 3 commits later :/

**Pull request closed without merging**